internal/controller/ngrok/cloudendpoint_controller.go

```diff
@@ -0,0 +1,342 @@
+	clep.Status.ID = ngrokClep.ID
+	return r.Client.Status().Update(ctx, clep)
+}
+
+// #region Helper Functions
+
+// findCloudEndpointForTrafficPolicy searches for any Cloud Endpoints CRs that have a reference to a particular Traffic Policy
+func (r *CloudEndpointReconciler) findCloudEndpointForTrafficPolicy(ctx context.Context, o client.Object) []ctrl.Request {
+	tp, ok := o.(*ngrokv1alpha1.NgrokTrafficPolicy)
+	if !ok {
+		return nil
+	}
+
+	// Use the index to find CloudEndpoints that reference this TrafficPolicy
+	var cloudEndpointList ngrokv1alpha1.CloudEndpointList
+	if err := r.Client.List(ctx, &cloudEndpointList,
+		client.InNamespace(tp.Namespace),
+		client.MatchingFields{trafficPolicyNameIndex: tp.Name}); err != nil {
+		r.Log.Error(err, "failed to list CloudEndpoints using index")
+		return nil
+	}
+
+	// Collect the requests for matching CloudEndpoints
+	var requests []ctrl.Request
+	for _, clep := range cloudEndpointList.Items {
+		requests = append(requests, ctrl.Request{
+			NamespacedName: client.ObjectKey{
+				Name:      clep.Name,
+				Namespace: clep.Namespace,
+			},
+		})
+	}
+
+	return requests
+}
+
+// findTrafficPolicy fetches the TrafficPolicy CRD from the API server and returns the JSON policy as a string
+func (r *CloudEndpointReconciler) findTrafficPolicy(ctx context.Context, tpName, tpNamespace string) (string, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("name", tpName, "namespace", tpNamespace)
+
+	// Create a TrafficPolicy object to store the fetched result
+	tp := &ngrokv1alpha1.NgrokTrafficPolicy{}
+	key := client.ObjectKey{Name: tpName, Namespace: tpNamespace}
+
+	// Attempt to get the TrafficPolicy from the API server
+	if err := r.Client.Get(ctx, key, tp); err != nil {
+		log.Error(err, "failed to find TrafficPolicy CRD")
+		return "", err
+	}
+
+	// Convert the JSON policy to a string
+	policyBytes, err := tp.Spec.Policy.MarshalJSON()
+	if err != nil {
+		log.Error(err, "failed to marshal TrafficPolicy JSON")
+		return "", err
+	}
+
+	return string(policyBytes), nil
+}
+
+// ensureDomainExists checks if the Domain CRD exists, and if not, creates it.
+func (r *CloudEndpointReconciler) ensureDomainExists(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint) error {
+	domain := r.extractDomain(clep)
+	if domainEndsInReservedTLD(domain) {
+		// Skip creating the Domain CRD for reserved TLDs
+		return nil
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("domain", domain)
+
+	// Check if the Domain CRD already exists
+	domainObj := &ingressv1alpha1.Domain{}
+	err := r.Get(ctx, client.ObjectKey{Name: domain, Namespace: clep.Namespace}, domainObj)
+	if err == nil {
+		// Domain already exists
+		return nil
+	}
+	if client.IgnoreNotFound(err) != nil {
+		// Some other error occurred
+		log.Error(err, "failed to check Domain CRD existence")
+		return err
+	}
+
+	// Create the Domain CRD
+	newDomain := &ingressv1alpha1.Domain{
+		ObjectMeta: ctrl.ObjectMeta{
+			Name:      domain,
```

Comment:
I think you'll need to sanitize the `Name` against the `.`. We may or may not want to pull this into a utility function https://github.com/ngrok/ngrok-operator/blob/0003f003816f63087dba730b70978cf23797d302/internal/store/driver.go#L34-L36

Reply:
I have test code for this we can pull out somewhere else: https://github.com/ngrok/ngrok-operator/blob/main/internal/controller/bindings/endpointbinding_poller_test.go#L307-L309

Reply:
@hjkatz are you referring to https://github.com/ngrok/ngrok-operator/blob/4e3428ab4772d59d20c5da44f5fbbaeca10eabef/internal/controller/bindings/boundendpoint_poller_test.go#L298-L311 ?

Reply:
Yes, sorry I changed the name and merged to main recently.

Reply:
@jonstacks I think I'll add a simple function to `api/ingress/v1alpha1/domain_types.go` that takes a string and returns the sanitized name and consume it in both spots